test/ConfigManagerTest.cpp

```diff
@@ -0,0 +1,347 @@
+
+  // 'Compare' two configuration options
+  auto compareConfigurationOptions = []<typename T>(
+                                         const ad_utility::ConfigOption& a,
+                                         const ad_utility::ConfigOption& b) {
+    ASSERT_EQ(a.hasSetDereferencedVariablePointer(),
+              b.hasSetDereferencedVariablePointer());
+
+    if (a.hasSetDereferencedVariablePointer()) {
+      ASSERT_EQ(a.getValue<T>(), b.getValue<T>());
+      ASSERT_EQ(a.getValueAsString(), b.getValueAsString());
+    }
+  };
+
+  // Configuration options for testing.
+  int notUsed;
+
+  config.createConfigOption(
+      {"Shared_part", "Unique_part_1", "Sense_of_existence"}, "", &notUsed,
+      std::optional{42});
+  const ad_utility::ConfigOption& withDefault{
+      ad_utility::ConfigOption("Sense_of_existence", "", &notUsed, 42)};
+
+  config.createConfigOption(
+      {"Shared_part", "Unique_part_2", size_t{3}, "Sense_of_existence"}, "",
+      &notUsed);
+  const ad_utility::ConfigOption& withoutDefault{
+      ad_utility::ConfigOption("Sense_of_existence", "", &notUsed)};
+
+  // Where those two options added?
+  ASSERT_EQ(config.getConfigurationOptions().size(), 2);
+
+  compareConfigurationOptions.template operator()<int>(
+      withDefault, config.getConfigurationOptionByNestedKeys(
+                       {"Shared_part", "Unique_part_1", "Sense_of_existence"}));
+  compareConfigurationOptions.template operator()<int>(
+      withoutDefault,
+      config.getConfigurationOptionByNestedKeys(
+          {"Shared_part", "Unique_part_2", size_t{3}, "Sense_of_existence"}));
+
+  // Trying to get a configuration option, that does not exist, should cause
+  // an exception.
+  ASSERT_THROW(
+      config.getConfigurationOptionByNestedKeys({"Shared_part", "Getsbourgh"}),
+      ad_utility::NoConfigOptionFoundException);
+}
+
+/*
+The exceptions for adding configuration options.
+*/
+TEST(ConfigManagerTest, CreateConfigurationOptionExceptionTest) {
+  ad_utility::ConfigManager config{};
+
+  // Configuration options for testing.
+  int notUsed;
+  config.createConfigOption<int>(
+      {"Shared_part", "Unique_part_1", "Sense_of_existence"}, "", &notUsed, 42);
+
+  // Trying to add a configuration option with the same name at the same
+  // place, should cause an error.
+  ASSERT_THROW(config.createConfigOption<int>(
+      {"Shared_part", "Unique_part_1", "Sense_of_existence"}, "", &notUsed, 42);
+               , ad_utility::ConfigManagerOptionPathAlreadyinUseException);
+
+  /*
+  If the first key for the path given is a number, that should cause an
+  exception.
+  Reason: We want our `tree` (a `nlohmann::json` object) to be a json
+  object literal, so that user can easier find things. Ordering your options
+  by just giving them numbers, would be bad practice, so we should prevent it.
+  */
+  ASSERT_THROW(
+      config.createConfigOption<int>(
+          {size_t{0}, "Shared_part", "Sense_of_existence"}, "", &notUsed, 42);
```

Comment:
I am still for only allowing strings here, and then just passing a vector of strings as the argument, but we can talk about this.

Reply:
Did the wanted change.